connectome_interpreter/compress_paths.py

```diff
@@ -48,25 +49,78 @@ def compress_paths(inprop, step_number, threshold=0, output_threshold=1e-4, root
         >>> compressed_paths = compress_paths(inprop, step_number, threshold=0.1, output_threshold=0.01)
         >>> print(compressed_paths)
     """
-    steps_fast = []
+    steps_fast: List[csc_matrix] = []
     device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
 
-    inprop_tensor = torch.tensor(inprop.toarray(), device=device)
+    # check that step_number>0
+    if step_number < 1:
+        raise ValueError("step_number should be greater than 0")
+
+    inprop_tensor = torch.tensor(inprop.toarray())
+    size = inprop_tensor.size(dim=0)
+
+    chunks = math.ceil(size/chunkSize)
+
+    # print(
+    #     f'max memory before start: {torch.cuda.max_memory_allocated()/ 1024**2:.2f} MB')
 
     with torch.no_grad():
         for i in tqdm(range(step_number)):
+
             if i == 0:
                 out_tensor = inprop_tensor.clone()
             else:
-                out_tensor = torch.matmul(out_tensor, inprop_tensor)
+                # print(f'begin iteration {i}')
+                # print(torch.cuda.max_memory_allocated() / 1024**2)
+                out_tensor_new = torch.full(
+                    (size, size), 0.0, dtype=torch.float32)
+                colLow = 0
+                colHigh = chunkSize
+                for colChunk in range(chunks):  # iterate chunks colwise
+                    rowLow = 0
+                    rowHigh = chunkSize
+
+                    in_col = inprop_tensor[:, colLow:colHigh].to(
+                        device)  # shape: size x chunkSize; on GPU
+                    # print('after setting in_col')
+                    # print(torch.cuda.max_memory_allocated() / 1024**2)
+
+                    for rowChunk in range(chunks):  # iterate chunks rowwise
+                        torch.cuda.reset_peak_memory_stats()
```

Comment:
If this is for stats and analytics, it would make a bit more sense to move this after the next assignment. See the note on `del` later.

Reply:
Sorry just forgot to remove it. Removed now.